manifest.go

```diff
@@ -324,6 +325,78 @@ func ValidateProjectRoots(c *Ctx, m *Manifest, sm gps.SourceManager) error {
 	return valErr
 }
 
+// ValidatePackageRules ensure that there are no ineffectual package declarations
+// in the constraints, overrides, required, and ignored rules in the manifest.
+func ValidatePackageRules(c *Ctx, proj *Project, sm gps.SourceManager) (err error) {
+	directDeps, err := proj.DirectDependencies(sm)
+	if err != nil {
+		return err
+	}
+
+	ic := FindIneffectualConstraints(proj.Manifest, directDeps)
+	if len(ic) > 0 {
+		err = errInefficientRules
+		c.Err.Printf("The following constraints aren't imported:\n\n")
+		for _, pr := range ic {
+			c.Err.Println("  ", pr)
+		}
+		c.Err.Println()
+	}
+
+	ig := FindIneffectualIgnores(proj.Manifest, directDeps)
+	if len(ig) > 0 {
+		err = errInefficientRules
+		c.Err.Printf("The following ignored packages aren't imported:\n\n")
+		for _, pr := range ig {
+			c.Err.Println("  ", pr)
+		}
+		c.Err.Println()
+	}
+
+	return err
+}
+
+// FindIneffectualConstraints looks for constraints decleared in the project
+// manifest that aren't directly used in the project. Required packages are an
+// exception to this check.
+func FindIneffectualConstraints(m *Manifest, directDeps map[string]bool) []gps.ProjectRoot {
+	// Treat required packages as if they were direct dependencies
+	for pr, _ := range m.RequiredPackages() {
+		directDeps[string(pr)] = true
+	}
+
+	ineffectuals := make([]gps.ProjectRoot, 0)
+	// Check constraints
+	pc := m.DependencyConstraints()
+	for pr, _ := range pc {
+		if !directDeps[string(pr)] {
+			ineffectuals = append(ineffectuals, pr)
+		}
+	}
+
+	// Check overrides as well
```

Comment:
Let's have proper sentences with a period at the end for all the comments :) We are doing it as much as possible for all the new code.